developer_guides/firmware/async_messaging_best_practices.rst

```diff
@@ -0,0 +1,651 @@
+.. _async_messaging_best_practices:
+
+Async Messaging Best Practices
+##############################
+
+This section provides best practices and point out issues that developers should
+be aware of when using asynchronous messages in their components.
+
+Message Type IDs
+****************
+
+The only unique value that is guaranteed to not change is UUID. The message type
+IDs can change every time when a component is registering as a message
+consumer/producer. In fact it depends on initialization order of components and
+it may seem to be static. However it can change i.e. with addition of a new
+component to pipeline and the entire mechanism will stop working.
+
+The components must not hardcode this value and they must use a runtime value of
+component type ID returned during producer/consumer registration.
+
+.. _message_handling:
+
+Message Handling
+****************
+
+The asynchronous messages consumers must be prepared to handle the messages
+asynchronously. A message can be received while a component is processing data
+and it can lead to undefined behavior when internal state is changed
+immediately. The recommended way of handling asynchronous messages is to save a
+message and apply the changes during a next processing time i.e. at the
+beginning of processing.
+
+-  When a processing of asynchronous message is done at the beginning of
+   component processing (Process Data), only then component processing is
+   deterministic.
+
+If an asynchronous message cannot be processed immediately by a consumer, it is
+a good practice to implement a queue of incoming asynchronous messages in a
+component or at least count asynchronous messages since last processing to
+detect this kind of situation.
```

Comment:
I assume the sender cant block, but could the sender get an error in the receiver queue was full ?

Reply:
No status is expected to be returned from async message processing to the sender. Overall you may have multiple consumers of a single event and all of them may result in a different way on an async event. It is probably more critical for the consumer to be aware that it received some event that could not fill into the queue, but it is up to component implementation and it event handling routine.